# Modeling and Simulation in Python

Chapter 3: Explain

Copyright 2017 Allen Downey

License: [Creative Commons Attribution 4.0 International](https://creativecommons.org/licenses/by/4.0)


In [1]:
# If you want the figures to appear in the notebook, 
# and you want to interact with them, use
# %matplotlib notebook

# If you want the figures to appear in the notebook, 
# and you don't want to interact with them, use
# %matplotlib inline

# If you want the figures to appear in separate windows, use
# %matplotlib qt5

# To switch from one to another, you have to select Kernel->Restart

%matplotlib inline

from modsim import *

Pandas is a module that provides tools for reading and processing data.  The `read_html` reads a web page from a file or the Internet and creates one DataFrame for each table on the page.

In [2]:
from pandas import read_html

The data directory contains a downloaded copy of https://en.wikipedia.org/wiki/World_population_estimates

In [3]:
filename = 'data/World_population_estimates.html'
tables = read_html(filename, header=0, index_col=0, decimal='M')

`tables` is a sequence of DataFrame objects.  We can select the DataFrame we want using the bracket operator.  The tables are numbered from 0, so `table2` is actually the third table on the page.

`head` selects the header and the first five rows.

In [4]:
table2 = tables[2]
table2.head()

United States Census Bureau (2015)[18]  \
Year                                           
1950                              2557628654   
1951                              2594939877   
1952                              2636772306   
1953                              2682053389   
1954                              2730228104   

      Population Reference Bureau (1973–2015)[6]  \
Year                                               
1950                                2.516000e+09   
1951                                         NaN   
1952                                         NaN   
1953                                         NaN   
1954                                         NaN   

      United Nations Department of Economic and Social Affairs (2015)[7]  \
Year                                                                       
1950                                         2525149000                    
1951                                         2572850917                    
1952                                         2619292068                    
1953                                         2665865392                    
1954                                         2713172027                    

      Maddison (2008)[8]  HYDE (2007)[15]  Tanton (1994)[9]  \
Year                                                          
1950        2.544000e+09     2.527960e+09      2.400000e+09   
1951        2.571663e+09              NaN               NaN   
1952        2.617949e+09              NaN               NaN   
1953        2.665959e+09              NaN               NaN   
1954        2.716927e+09              NaN               NaN   

      Biraben (1980)[10]  McEvedy & Jones (1978)[11]  Thomlinson (1975)[12]  \
Year                                                                          
1950        2.527000e+09                2.500000e+09           2.400000e+09   
1951                 NaN                         NaN                    NaN   
1952                 NaN                         NaN                    NaN   
1953                 NaN                         NaN                    NaN   
1954                 NaN                         NaN                    NaN   

     Durand (1974)[13]  Clark (1967)[14]  
Year                                      
1950               NaN      2.486000e+09  
1951               NaN               NaN  
1952               NaN               NaN  
1953               NaN               NaN  
1954               NaN               NaN

`tail` selects the last five rows.

In [5]:
table2.tail()

United States Census Bureau (2015)[18]  \
Year                                           
2011                              6944055583   
2012                              7022349283   
2013                              7101027895   
2014                              7178722893   
2015                              7256490011   

      Population Reference Bureau (1973–2015)[6]  \
Year                                               
2011                                6.986951e+09   
2012                                7.057075e+09   
2013                                7.136796e+09   
2014                                7.238184e+09   
2015                                7.336435e+09   

      United Nations Department of Economic and Social Affairs (2015)[7]  \
Year                                                                       
2011                                         6997998760                    
2012                                         7080072417                    
2013                                         7162119434                    
2014                                         7243784000                    
2015                                         7349472000                    

      Maddison (2008)[8]  HYDE (2007)[15]  Tanton (1994)[9]  \
Year                                                          
2011                 NaN              NaN               NaN   
2012                 NaN              NaN               NaN   
2013                 NaN              NaN               NaN   
2014                 NaN              NaN               NaN   
2015                 NaN              NaN               NaN   

      Biraben (1980)[10]  McEvedy & Jones (1978)[11]  Thomlinson (1975)[12]  \
Year                                                                          
2011                 NaN                         NaN                    NaN   
2012                 NaN                         NaN                    NaN   
2013                 NaN                         NaN                    NaN   
2014                 NaN                         NaN                    NaN   
2015                 NaN                         NaN                    NaN   

     Durand (1974)[13]  Clark (1967)[14]  
Year                                      
2011               NaN               NaN  
2012               NaN               NaN  
2013               NaN               NaN  
2014               NaN               NaN  
2015               NaN               NaN

Long column names are awkard to work with, but we can replace them with abbreviated names.

In [6]:
table2.columns = ['census', 'prb', 'un', 'maddison', 
                  'hyde', 'tanton', 'biraben', 'mj', 
                  'thomlinson', 'durand', 'clark']

Here's what the DataFrame looks like now.  

Some of the values use scientific notation; for example, `2.544000e+09` is shorthand for $2.544 \cdot 10^9$ or 2.544 billion.

`NaN` is a special value that indicates missing data.

In [7]:
table2

census           prb          un      maddison          hyde  \
Year                                                                     
1950  2557628654  2.516000e+09  2525149000  2.544000e+09  2.527960e+09   
1951  2594939877           NaN  2572850917  2.571663e+09           NaN   
1952  2636772306           NaN  2619292068  2.617949e+09           NaN   
1953  2682053389           NaN  2665865392  2.665959e+09           NaN   
1954  2730228104           NaN  2713172027  2.716927e+09           NaN   
1955  2782098943           NaN  2761650981  2.769074e+09           NaN   
1956  2835299673           NaN  2811572031  2.822502e+09           NaN   
1957  2891349717           NaN  2863042795  2.879934e+09           NaN   
1958  2948137248           NaN  2916030167  2.939254e+09           NaN   
1959  3000716593           NaN  2970395814  2.995909e+09           NaN   
1960  3043001508           NaN  3026002942  3.041507e+09  3.042000e+09   
1961  3083966929           NaN  3082830266  3.082161e+09           NaN   
1962  3140093217           NaN  3141071531  3.135787e+09           NaN   
1963  3209827882           NaN  3201178277  3.201354e+09           NaN   
1964  3281201306           NaN  3263738832  3.266477e+09           NaN   
1965  3350425793           NaN  3329122479  3.333138e+09           NaN   
1966  3420677923           NaN  3397475247  3.402224e+09           NaN   
1967  3490333715           NaN  3468521724  3.471464e+09           NaN   
1968  3562313822           NaN  3541674891  3.543086e+09           NaN   
1969  3637159050           NaN  3616108749  3.615743e+09           NaN   
1970  3712697742           NaN  3691172616  3.691157e+09  3.710000e+09   
1971  3790326948           NaN  3766754345  3.769818e+09           NaN   
1972  3866568653           NaN  3842873611  3.846499e+09           NaN   
1973  3942096442           NaN  3919182332  3.922793e+09  3.923000e+09   
1974  4016608813           NaN  3995304922  3.997677e+09           NaN   
1975  4089083233           NaN  4071020434  4.070671e+09           NaN   
1976  4160185010           NaN  4146135850  4.141445e+09           NaN   
1977  4232084578           NaN  4220816737  4.213539e+09           NaN   
1978  4304105753           NaN  4295664825  4.286317e+09           NaN   
1979  4379013942           NaN  4371527871  4.363144e+09           NaN   
...          ...           ...         ...           ...           ...   
1986  4940571232           NaN  4953376710  4.920968e+09           NaN   
1987  5027200492           NaN  5045315871  5.006672e+09           NaN   
1988  5114557167           NaN  5138214688  5.093306e+09           NaN   
1989  5201440110           NaN  5230000000  5.180540e+09           NaN   
1990  5288955934           NaN  5320816667  5.269029e+09  5.308000e+09   
1991  5371585922           NaN  5408908724  5.351922e+09           NaN   
1992  5456136278           NaN  5494899570  5.435722e+09           NaN   
1993  5538268316           NaN  5578865109  5.518127e+09           NaN   
1994  5618682132           NaN  5661086346  5.599396e+09           NaN   
1995  5699202985  5.760000e+09  5741822412  5.681575e+09           NaN   
1996  5779440593           NaN  5821016750  5.762212e+09           NaN   
1997  5857972543  5.840000e+09  5898688337  5.842122e+09           NaN   
1998  5935213248           NaN  5975303657  5.921366e+09           NaN   
1999  6012074922           NaN  6051478010  5.999622e+09           NaN   
2000  6088571383  6.067000e+09  6127700428  6.076558e+09  6.145000e+09   
2001  6165219247  6.137000e+09  6204147026  6.154791e+09           NaN   
2002  6242016348  6.215000e+09  6280853817  6.231704e+09           NaN   
2003  6318590956  6.314000e+09  6357991749  6.308364e+09           NaN   
2004  6395699509  6.396000e+09  6435705595  6.374056e+09           NaN   
2005  6473044732  6.477000e+09  6514094605  6.462987e+09           NaN   
2006  6551263534  6.555000e+09  6593227977  6.540214e+09           NaN   
2007  66299137

We can use dot notatio to select a column from a DataFrame.  The result is a Series.

In [8]:
census = table2.census
census

Year
1950    2557628654
1951    2594939877
1952    2636772306
1953    2682053389
1954    2730228104
1955    2782098943
1956    2835299673
1957    2891349717
1958    2948137248
1959    3000716593
1960    3043001508
1961    3083966929
1962    3140093217
1963    3209827882
1964    3281201306
1965    3350425793
1966    3420677923
1967    3490333715
1968    3562313822
1969    3637159050
1970    3712697742
1971    3790326948
1972    3866568653
1973    3942096442
1974    4016608813
1975    4089083233
1976    4160185010
1977    4232084578
1978    4304105753
1979    4379013942
           ...    
1986    4940571232
1987    5027200492
1988    5114557167
1989    5201440110
1990    5288955934
1991    5371585922
1992    5456136278
1993    5538268316
1994    5618682132
1995    5699202985
1996    5779440593
1997    5857972543
1998    5935213248
1999    6012074922
2000    6088571383
2001    6165219247
2002    6242016348
2003    6318590956
2004    6395699509
2005    6473044732
2006    6551263534
2007   

A Series object has two parts, `values` and `index`.

The `values` part is an array.

In [9]:
census.values

array([2557628654, 2594939877, 2636772306, 2682053389, 2730228104,
       2782098943, 2835299673, 2891349717, 2948137248, 3000716593,
       3043001508, 3083966929, 3140093217, 3209827882, 3281201306,
       3350425793, 3420677923, 3490333715, 3562313822, 3637159050,
       3712697742, 3790326948, 3866568653, 3942096442, 4016608813,
       4089083233, 4160185010, 4232084578, 4304105753, 4379013942,
       4451362735, 4534410125, 4614566561, 4695736743, 4774569391,
       4856462699, 4940571232, 5027200492, 5114557167, 5201440110,
       5288955934, 5371585922, 5456136278, 5538268316, 5618682132,
       5699202985, 5779440593, 5857972543, 5935213248, 6012074922,
       6088571383, 6165219247, 6242016348, 6318590956, 6395699509,
       6473044732, 6551263534, 6629913759, 6709049780, 6788214394,
       6866332358, 6944055583, 7022349283, 7101027895, 7178722893,
       7256490011], dtype=int64)

The `index` part is yet another kind of object, an `Int64Index`.

In [10]:
census.index

Int64Index([1950, 1951, 1952, 1953, 1954, 1955, 1956, 1957, 1958, 1959, 1960,
            1961, 1962, 1963, 1964, 1965, 1966, 1967, 1968, 1969, 1970, 1971,
            1972, 1973, 1974, 1975, 1976, 1977, 1978, 1979, 1980, 1981, 1982,
            1983, 1984, 1985, 1986, 1987, 1988, 1989, 1990, 1991, 1992, 1993,
            1994, 1995, 1996, 1997, 1998, 1999, 2000, 2001, 2002, 2003, 2004,
            2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015],
           dtype='int64', name='Year')

If you ever wonder what kind of object a variable refers to, you can use the `type` function.

The result indicates what type the object is, and the module where that type is defined.

DataFrame, Series, and Int64Index are defined by Pandas.

array is defined by NumPy.

In [11]:
type(table2)

pandas.core.frame.DataFrame

In [12]:
type(census)

pandas.core.series.Series

In [13]:
type(census.index)

pandas.core.indexes.numeric.Int64Index

In [14]:
type(census.values)

numpy.ndarray

This function plots the estimates generated by the US Censis and UN DESA, and labels the axes.

`1e9` is scientific notation for $1 \cdot 10^9$ or 1 billion.

In [15]:
def plot_estimates(table):
    """Plot world population estimates.
    
    table: DataFrame with columns 'un' and 'census'
    """
    un = table.un / 1e9
    census = table.census / 1e9
    
    plot(census, ':', color='darkblue', label='US Census')
    plot(un, '--', color='green', label='UN DESA')
    
    decorate(xlabel='Year',
             ylabel='World population (billion)')

Now we can plot the estimates.

In [17]:
newfig()
plot_estimates(table2)
savefig('chap03-fig01.pdf')

From here on, we will work in units of billions.

In [18]:
un = table2.un / 1e9

In [19]:
census = table2.census / 1e9

This expression computes the elementwise differences between the series, then divides through by the UN value to produce relative errors, then finds the largest element.

So the largest relative error between the estimates is about 1.3%.

In [20]:
max(abs(census - un) / un) * 100

**Exercise:** Break down that expression into smaller steps and display the intermediate results, to make sure you understand how it works.

Where in the series is the largest relative error between the two estimates, near the beginning or the end?

When I computed relative errors, I used `un` as the denominator.  But that was an arbitraty choice.  What happens if we use `census` instead?  How much difference does it make.

In [21]:
# Solution goes here

In [22]:
# Solution goes here

In [23]:
# Solution goes here

In [24]:
# Solution goes here

### Constant growth

We can select an element from a series using bracket notation and one of the elements from the index.  Here's the first element:

In [25]:
census[1950]

And the last element.

In [26]:
census[2015]

But we can get the first and last years from the index itself:

In [27]:
first_year = census.index[0]
last_year = census.index[-1]
first_year, last_year

And use them to look up the first and last elements.

Then we can compute the average annual growth in billions of people per year.

In [28]:
total_growth = census[last_year] - census[first_year]
elapsed_time = last_year - first_year
annual_growth = total_growth / elapsed_time
annual_growth

Now let's create a `TimeSeries` to contain values generated by a linear growth model.

In [29]:
results = TimeSeries()

Initially the Series is empty, but we can initialize it so the starting value, in 1950, is the 1950 population estimated by the US Census.

In [30]:
results[1950] = census[1950]
results

After that, the population in the model grows by a constant amount each year.

In [31]:
for t in linrange(1950, 2015):
    results[t+1] = results[t] + annual_growth

Here's what the results looks like, compared to the actual data.

In [32]:
newfig()
plot_estimates(table2)
plot(results, '--', color='gray', label='model')
decorate(xlabel='Year', ylabel='World population (billion)')
savefig('chap03-fig02.pdf')

The model fits the data pretty well after 1990, but not so well before.

**Exercise:**  Try fitting the model using data from 1970 to the present, and see if that does a better job.

Hint: Copy the code from above and make a few changes.

Make sure your model starts in 1950, even though the estimated annual growth is based on later data.  You might have to shift the first value in the series up or down to match the data.

In [33]:
# Solution goes here

### Now with system objects

We can rewrite the code from the previous section using system objects.

In [34]:
t0 = census.index[0]
t_end = census.index[-1]
total_growth = census[t_end] - census[t0]
elapsed_time = t_end - t0
annual_growth = total_growth / elapsed_time

Here's the system object.

In [35]:
system = System(t0=t0, 
                t_end=t_end,
                p0=census[t0],
                annual_growth=annual_growth)

And we can encapsulate the code tha runs the model in a function that stores the resulting Series as a new system variable.

In [36]:
def run_simulation1(system):
    """Runs the constant growth model.
    
    Adds TimeSeries to `system` as `results`.
    
    system: system object
    """
    results = TimeSeries()
    results[system.t0] = system.p0
    for t in linrange(system.t0, system.t_end):
        results[t+1] = results[t] + system.annual_growth
    system.results = results

We can also encapsulate the code that plots the results.

In [37]:
def plot_results(system, title=None):
    """Plot the estimates and the model.
    
    system: System object with `results`
    """
    newfig()
    plot_estimates(table2)
    plot(system.results, '--', color='gray', label='model')
    decorate(xlabel='Year', 
             ylabel='World population (billion)',
             title=title)

Here's how we run it.

In [38]:
run_simulation1(system)
plot_results(system, title='Constant growth model')

`plot_results` uses `decorate`, which takes parameters that specify the title of the figure, labels for the $x$ and $y$ axis, and limits for the axes.  To read the documentation of `decorate`, run the cells below.

In [39]:
help(decorate)

In [40]:
plt.getp(plt.gca())

**Exercise:** The constant growth model doesn't make a lot of sense, because it seems like the number of deaths and births should depend on the size of the population.  As a small improvement, let's write a version of `run_simulation1` where the number of deaths is proportional to the size of the population, but the number of births is constant.  This model doesn't make a lot of sense, either, but it's a good exercise.

Write a function called `run_simulation1b` that implements a model where the number of births is constant, but the number of deaths is proportional to the current size of the population.  Set the death rate to `0.01`, which means that 1% of the population dies each year; then choose the number of annual births to make the model fit the data as well as you can.

Hint: It probably won't fit very well.

In [41]:
# Solution goes here

In [42]:
# Solution goes here

## Proportional death, proportional birth

Now let's get to a more realistic model where the number of births and deaths is proportional to the current population.

In [43]:
def run_simulation2(system):
    """Runs the constant growth model.
    
    Adds TimeSeries to `system` as `results`.
    
    system: system object
    """
    results = TimeSeries()
    results[system.t0] = system.p0
    for t in linrange(system.t0, system.t_end):
        births = system.birth_rate * results[t]
        deaths = system.death_rate * results[t]
        results[t+1] = results[t] + births - deaths
    system.results = results

I kept the death rate at 1% and chose the birth rate to fit the data.

In [44]:
system.death_rate = 0.01
system.birth_rate = 0.027

Here's what it looks like.

In [45]:
run_simulation2(system)
plot_results(system, title='Proportional model')
savefig('chap03-fig03.pdf')

The model fits the data pretty well for the first 20 years, but not so well after that.

**Exercise:** In this implementation, we compute the number of deaths and births separately, but since they are both proportional to the current population, we can combine them.

Write a function called `run_simulation2b` that implements a model with a single parameter, `alpha`, that represents the net growth rate, which is the difference between the birth and death rates.  For example, if `alpha=0.01`, the population should grow by 1% per year.

Choose the value of `alpha` that fits the data best.

In [46]:
# Solution goes here

In [47]:
# Solution goes here

### Factoring out the update function

The functions that run the model all look the same except the body of the loop.  So we can factor that part out into a function.

In [48]:
def update_func1(pop, t, system):
    """Compute the population next year.
    
    pop: current population
    t: current year
    system: system object containing parameters of the model
    
    returns: population next year
    """
    births = system.birth_rate * pop
    deaths = system.death_rate * pop
    return pop + births - deaths

Now the name `update_func1` refers to a function object.

In [49]:
update_func1

Which we can confirm by checking its type.

In [50]:
type(update_func1)

`run_simulation` takes the update function as a parameter and calls it just like any other function.

In [51]:
def run_simulation(system, update_func):
    """Simulate the system using any update function.
    
    Adds TimeSeries to `system` as `results`.

    system: System object
    update_func: function that computes the population next year
    """
    results = TimeSeries()
    results[system.t0] = system.p0
    for t in linrange(system.t0, system.t_end):
        results[t+1] = update_func(results[t], t, system)
    system.results = results

Here's how we use it.

In [52]:
run_simulation(system, update_func1)

Remember not to put parentheses after `update_func1`.  What happens if you try?

In [53]:
plot_results(system, title='Proportional model, factored')

**Exercise:** When you run `run_simulation`, it runs `update_func1` once for each year between `t0` and `t_end`.  To see that for yourself, add a print statement at the beginning of `update_func1` that prints the values of `t` and `pop`, then run `run_simulation` again.

### Combining birth and death

Since births and deaths get added up, we don't have to compute them separately.  We can combine the birth and death rates into a single net growth rate.

In [54]:
def update_func1b(pop, t, system):
    """Compute the population next year.
    
    pop: current population
    t: current year
    system: system object containing parameters of the model
    
    returns: population next year
    """
    net_growth = system.alpha  * pop
    return pop + net_growth

Here's how it works:

In [55]:
system.alpha = system.birth_rate - system.death_rate

run_simulation(system, update_func1b)
plot_results(system, title='Proportional model, combined birth and death')

**Exercise:** Maybe the reason the proportional model doesn't work very well is that the growth rate, `alpha`, might be changing over time.  So let's try a model with different growth rates before and after 1980 (as an arbitrary choice).

Write a function called `update_func1c` that takes `pop`, `t`, and `system` as parameters.  The system object, `system`, should contains two parameters: the growth rate before 1980, `alpha1`, and the growth rate after 1980, `alpha2`.  It should compute and return the simulated population one year later.

Note: Don't forget the `return` statement.

In [56]:
# Solution goes here

In [57]:
# Solution goes here

### Quadratic growth

Here's the implementation of the quadratic growth model.

In [58]:
def update_func2(pop, t, system):
    """Compute the population next year.
    
    pop: current population
    t: current year
    system: system object containing parameters of the model
    
    returns: population next year
    """
    net_growth = system.alpha * pop + system.beta * pop**2
    return pop + net_growth

And here are the results.  Can you find values for the parameters that make the model fit better?

In [59]:
system.alpha = 0.025
system.beta = -0.0018

run_simulation(system, update_func2)
plot_results(system, title='Quadratic model')
savefig('chap03-fig04.pdf')

To understand the quadratic model better, let's plot net growth as a function of population.

In [60]:
pop_array = linspace(0.001, 15, 100)
net_growth_array = system.alpha * pop_array + system.beta * pop_array**2

Here's what it looks like.  Remember that the x axis is population now, not time.

The function `sns.set` sets the style for the plots.  I added a grid to this one to make it easier to read.

In [61]:
newfig()
sns.set(style='whitegrid', font_scale=1.5)
plot(pop_array, net_growth_array, '-')
decorate(xlabel='Population (billions)',
         ylabel='Net growth (billions)',
         legend=False)
savefig('chap03-fig05.pdf')

Using `sns.set` to reset the plot style.

In [62]:
sns.set(style='white', font_scale=1.5)

In the book we found that the net growth is 0 when the population is $-\alpha/\beta$:

In [63]:
-system.alpha / system.beta

This is the equilibrium the population tends toward.

**Exercise:** In the book, I presented a different way to parameterize the quadratic model:

$ \Delta p = r p (1 - p / K) $

where $r=\alpha$ and $K=-\alpha/\beta$.  Write a version of `update_func2` that implements this version of the model.  Test it by computing system variables `r` and `K` equivalent to `alpha` and `beta`, and confirm that you get the same results. 

In [64]:
# Solution goes here

In [65]:
# Solution goes here

In [66]:
# Solution goes here

**Exercise:** On the Wikipedia page about world population estimates, the first table contains estimates for prehistoric populations.  The following cells process this table and plot some of the results.

Select `table1`, which is the second table on the page.

In [67]:
table1 = tables[1]
table1.head()

Not all agencies and researchers provided estimates for the same dates.  Again `NaN` is the special value that indicates missing data.

In [68]:
table1.tail()

Some of the estimates are in a form we can't read as numbers.  We could clean them up by hand, but for simplicity I'll replace any value that has an `M` in it with `NaN`.

In [69]:
table1.replace('M', np.nan, regex=True, inplace=True)

Again, we'll replace the long column names with more convenient abbreviations.

In [70]:
table1.columns = ['prb', 'un', 'maddison', 'hyde', 'tanton', 
                  'biraben', 'mj', 'thomlinson', 'durand', 'clark']

This function plots selected estimates.

In [71]:
def plot_prehistory(table):
    """Plots population estimates.
    
    table: DataFrame
    """
    plot(table.prb, 'ro', label='PRB')
    plot(table.un, 'co', label='UN')
    plot(table.hyde, 'yo', label='HYDE')
    plot(table.tanton, 'go', label='Tanton')
    plot(table.biraben, 'bo', label='Biraben')
    plot(table.mj, 'mo', label='McEvedy & Jones')

Here are the results.  Notice that we are working in millions now, not billions.

In [72]:
newfig()
plot_prehistory(table1)
decorate(xlabel='Year', 
         ylabel='World population (millions)',
         title='Prehistorical population estimates')

We can use `xlim` to zoom in on everything after Year 0.

In [73]:
newfig()
plot_prehistory(table1)
decorate(xlim=[0, 2000], xlabel='Year', 
         ylabel='World population (millions)',
         title='Prehistorical population estimates')

See if you can find a model that fits these data well from Year -1000 to 1940, or from Year 0 to 1940.

How well does your best model predict actual population growth from 1950 to the present?

In [74]:
# Solution goes here

In [75]:
# Solution goes here

In [76]:
# Solution goes here

### Disfunctions

In [78]:
def carrying_capacity(system):
    K = -system.alpha / system.beta
    return K
    
sys1 = System(alpha=0.025, beta=-0.0018)
pop = carrying_capacity(sys1)
print(pop)

In [80]:
def carrying_capacity():
    K = -system.alpha / system.beta
    return K
    
system = System(alpha=0.025, beta=-0.0018)
pop = carrying_capacity()
print(pop)

In [81]:
def carrying_capacity(system):
    system = System(alpha=0.025, beta=-0.0018)
    K = -system.alpha / system.beta
    return K
    
sys1 = System(alpha=0.025, beta=-0.0018)
pop = carrying_capacity(sys1)
print(pop)

In [82]:
def carrying_capacity(system):
    K = -system.alpha / system.beta
    
sys1 = System(alpha=0.025, beta=-0.0018)
pop = carrying_capacity(sys1)
print(pop)

In [83]:
def carrying_capacity(system):
    K = -system.alpha / system.beta
    return K
    
sys1 = System(alpha=0.025, beta=-0.0018)
carrying_capacity(sys1)
print(K)